--- a/src/gba/savedata.c
+++ b/src/gba/savedata.c
@@ -69,13 +69,16 @@
 }
 
 void GBASavedataDeinit(struct GBASavedata* savedata) {
+	mLOG(GBA_SAVE, WARN, "ML3DTRACE deinit-in sd=%p type=%i data=%p vf=%p", savedata, savedata->type, savedata->data, savedata->vf);
 	if (savedata->vf) {
 		size_t size = GBASavedataSize(savedata);
 		if (savedata->data) {
 			savedata->vf->unmap(savedata->vf, savedata->data, size);
 		}
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE deinit-unmap sd=%p data=%p", savedata, savedata->data);
 		savedata->vf = NULL;
 	} else {
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE deinit-free sd=%p type=%i data=%p", savedata, savedata->type, savedata->data);
 		switch (savedata->type) {
 		case GBA_SAVEDATA_SRAM:
 			mappedMemoryFree(savedata->data, GBA_SIZE_SRAM);
@@ -226,6 +229,7 @@
 }
 
 void GBASavedataForceType(struct GBASavedata* savedata, enum GBASavedataType type) {
+	mLOG(GBA_SAVE, WARN, "ML3DTRACE forcetype sd=%p from=%i to=%i data=%p", savedata, savedata->type, type, savedata->data);
 	if (savedata->type == type) {
 		return;
 	}
@@ -265,6 +269,7 @@
 }
 
 void GBASavedataInitFlash(struct GBASavedata* savedata) {
+	mLOG(GBA_SAVE, WARN, "ML3DTRACE flash-in sd=%p type=%i data=%p vf=%p", savedata, savedata->type, savedata->data, savedata->vf);
 	if (savedata->type == GBA_SAVEDATA_AUTODETECT) {
 		savedata->type = GBA_SAVEDATA_FLASH512;
 	}
@@ -280,6 +285,7 @@
 	if (!savedata->vf) {
 		end = 0;
 		savedata->data = anonymousMemoryMap(GBA_SIZE_FLASH1M);
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE flash-alloc sd=%p data=%p size=%i", savedata, savedata->data, (int) GBA_SIZE_FLASH1M);
 	} else {
 		end = savedata->vf->size(savedata->vf);
 		if (end < flashSize) {
@@ -311,6 +317,7 @@
 	if (!savedata->vf) {
 		end = 0;
 		savedata->data = anonymousMemoryMap(GBA_SIZE_EEPROM);
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE eeprom-alloc sd=%p data=%p size=%i", savedata, savedata->data, (int) GBA_SIZE_EEPROM);
 	} else {
 		end = savedata->vf->size(savedata->vf);
 		if (end < eepromSize) {
@@ -335,6 +342,7 @@
 	if (!savedata->vf) {
 		end = 0;
 		savedata->data = anonymousMemoryMap(GBA_SIZE_SRAM);
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE sram-alloc sd=%p data=%p size=%i", savedata, savedata->data, (int) GBA_SIZE_SRAM);
 	} else {
 		end = savedata->vf->size(savedata->vf);
 		if (end < GBA_SIZE_SRAM) {
@@ -360,6 +368,7 @@
 	if (!savedata->vf) {
 		end = 0;
 		savedata->data = anonymousMemoryMap(GBA_SIZE_SRAM512);
+		mLOG(GBA_SAVE, WARN, "ML3DTRACE sram512-alloc sd=%p data=%p size=%i", savedata, savedata->data, (int) GBA_SIZE_SRAM512);
 	} else {
 		end = savedata->vf->size(savedata->vf);
 		if (end < GBA_SIZE_SRAM512) {
