--- a/src/platform/posix/memory.c
+++ b/src/platform/posix/memory.c
@@ -4,6 +4,8 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 #include <mgba-util/memory.h>
+
+#include <stdio.h>
 
 #ifndef DISABLE_ANON_MMAP
 #ifdef __SANITIZE_ADDRESS__
@@ -19,18 +21,24 @@
 #include <sys/mman.h>
 
 void* anonymousMemoryMap(size_t size) {
-	return mmap(0, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+	void* ptr = mmap(0, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+	printf("ML3DMAP alloc mmap ptr=%p size=%u\n", ptr, (unsigned) size);
+	return ptr;
 }
 
 void mappedMemoryFree(void* memory, size_t size) {
+	printf("ML3DMAP free munmap ptr=%p size=%u\n", memory, (unsigned) size);
 	munmap(memory, size);
 }
 #else
 void* anonymousMemoryMap(size_t size) {
-	return calloc(1, size);
+	void* ptr = calloc(1, size);
+	printf("ML3DMAP alloc calloc ptr=%p size=%u\n", ptr, (unsigned) size);
+	return ptr;
 }
 
 void mappedMemoryFree(void* memory, size_t size) {
+	printf("ML3DMAP free free ptr=%p size=%u\n", memory, (unsigned) size);
 	UNUSED(size);
 	free(memory);
 }
